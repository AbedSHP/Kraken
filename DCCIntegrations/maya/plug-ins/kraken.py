import sys, json

from PySide import QtGui

import kraken.ui
from kraken.ui.kraken_ui import KrakenUI

from maya import cmds
# import maya.OpenMaya as OpenMaya
import maya.OpenMayaUI as OpenMayaUI
import maya.OpenMayaMPx as OpenMayaMPx

try:
  # Maya 2013 with custom pyside build
  import PySide.shiboken as shiboken
except:
  # Maya 2014 and higher
  import shiboken



# Command
class OpenKrakenEditorCommand(OpenMayaMPx.MPxCommand):
  def __init__(self):
    OpenMayaMPx.MPxCommand.__init__(self)

  # Invoked when the command is run.
  def doIt(self,argList):
    parent = OpenMayaUI.MQtUtil.mainWindow()
    parent = shiboken.wrapInstance(long(parent), QtGui.QWidget)
    dfgApp = KrakenUI(parent)

  # Creator
  @staticmethod
  def creator():
    return OpenMayaMPx.asMPxPtr( OpenKrakenEditorCommand() )


class KrakenUndoableCmd(OpenMayaMPx.MPxCommand):

  def __init__(self):
    OpenMayaMPx.MPxCommand.__init__(self)

  def isUndoable(self):
    return True

  def doIt(self, argList):
    return 0

  def redoIt(self):
    SingletonHost.getController().redo()
    return 0

  def undoIt(self):
    SingletonHost.getController().undo()
    return 0

  @staticmethod
  def creator():
    return OpenMayaMPx.asMPxPtr( KrakenUndoableCmd() )


# Initialize the script plug-in
def initializePlugin(mobject):
  mplugin = OpenMayaMPx.MFnPlugin(mobject)

  try:
    mplugin.registerCommand( 'openKrakenEditor', OpenKrakenEditorCommand.creator )
  except:
    sys.stderr.write( 'Failed to register DFG commands:openKrakenEditor' )
    raise

  try:
    mplugin.registerCommand( 'krakenUndoableCmd', KrakenUndoableCmd.creator )
  except:
    sys.stderr.write( 'Failed to register DFG commands:krakenUndoableCmd' )
    raise

# Uninitialize the script plug-in
def uninitializePlugin(mobject):
  mplugin = OpenMayaMPx.MFnPlugin(mobject)

  try:
    mplugin.deregisterCommand( 'openKrakenEditor' )
  except:
    sys.stderr.write( 'Failed to unregister command: openKrakenEditor' )


  try:
    mplugin.deregisterCommand( 'krakenUndoableCmd' )
  except:
    sys.stderr.write( 'Failed to unregister command: krakenUndoableCmd' )
