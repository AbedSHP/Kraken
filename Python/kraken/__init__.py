"""Kraken Framework."""

__all__ = ['core', 'examples', 'helpers', 'plugins', 'gui']