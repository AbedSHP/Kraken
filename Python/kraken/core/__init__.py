"""Kraken Core."""

VERSION_MAJOR = 0
VERSION_MINOR = 0
VERSION_BUILD = 1
VERSION_SUFFIX = "alpha"


def getVersion():
    """Contatenates the version globals and returns the current version of
    Kraken.

    Return:
    String, current version of Kraken.

    """

    versionString = str(VERSION_MAJOR) + "." + str(VERSION_MINOR) + "." + str(VERSION_BUILD)
    if VERSION_SUFFIX:
        versionString = versionString + "-" + VERSION_SUFFIX

    return versionString
