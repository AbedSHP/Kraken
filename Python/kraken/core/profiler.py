"""Kraken - objects.profiler module.

Classes:
Profiler - Profiler Object.

"""
import time




class Profiler(object):
    """Kraken profiler object for debugging performance issues."""

    __instance = None

    class ProfilerItem(object):

        def __init__(self, label):
            super(Profiler.ProfilerItem, self).__init__()

            t = time.time()
            self.label = label
            self.start = t
            self.end = t
            self.children = []

        def addChild(self, item):
            self.children.append(item)

        def endProfiling(self):
            self.end = time.time()

    def __init__(self, label='Root'):
        super(Profiler, self).__init__()
        self.reset(label=label)

    def reset(self, label='Root'):
        """Resets the profiler for generating a new report

        Return:
        None

        """
        self.__roots = []
        self.__stack = []

    def push(self, label):
        """Adds a new child to the profiling tree and activates it.

        Return:
        None

        """

        item = Profiler.ProfilerItem(label)
        if len(self.__stack) == 0:
            self.__roots.append(item)
        else:
            self.__stack[-1].addChild(item)
        self.__stack.append(item)


    def pop(self):
        """Deactivates the current item in the tree and returns the profiler to the parent item

        Return:
        None

        """
        end = time.time()
        if len(self.__stack) == 0:
            raise Exception("Unable to close bracket. Pop has been called more times than push.") 
             
        self.__stack[-1].endProfiling()
        self.__stack.pop()
        
    def generateReport(self):
        """Returns a json object that encodes the data gathered during profiling.

        Return:
        the json object

        """
        if len(self.__stack) != 0:
            raise Exception("Profiler brackets not closed properly. pop must be called for every call to push. Pop needs to be called another " + str(len(self.__stack)) + " times") 
        report = {
            'tree': None
        }
        functions = {}
        def reportItem(item):
            itemReport  = {
                'label': item.label,
                'duration' : item.end - item.start
            }
            if item.label not in functions:
                functions[item.label] = itemReport['duration']
            else:
                functions[item.label] += itemReport['duration']

            for childItem in item.children:
                if 'children' not in itemReport:
                    itemReport['children'] = []
                itemReport['children'].append(reportItem(childItem))
            return itemReport

        report['tree'] = []
        for rootItem in self.__roots:
            report['tree'].append(reportItem(rootItem))
        return report

    @classmethod
    def getInstance(cls):
        """This class method returns the singleton instance for the Profiler

        Return:
        The singleton instance.

        """

        if cls.__instance is None:
            cls.__instance = Profiler()

        return cls.__instance
