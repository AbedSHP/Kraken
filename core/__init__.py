"""Kraken Framework."""

VERSION_MAJOR = 0
VERSION_MINOR = 0
VERSION_BUILD = 1
VERSION_SUFFIX = "beta"


def getVersion():
    versionString = str(VERSION_MAJOR) + "." + str(VERSION_MINOR) + "." + str(VERSION_BUILD)
    if VERSION_SUFFIX:
        versionString = versionString + "-" + VERSION_SUFFIX

    return versionString