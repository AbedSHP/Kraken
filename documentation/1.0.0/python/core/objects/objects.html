

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Objects &mdash; Kraken Rigging Framework 1.0.0 documentation</title>
  

  
  

  

  
  
    

  

  
  
    <link rel="stylesheet" href="../../../_static/css/theme.css" type="text/css" />
  

  
    <link rel="stylesheet" href="../../../_static/lowdown.css" type="text/css" />
  

  
    <link rel="top" title="Kraken Rigging Framework 1.0.0 documentation" href="../../../index.html"/>
        <link rel="up" title="Kraken Core" href="../core.html"/>
        <link rel="next" title="Attributes" href="attributes/attributes.html"/>
        <link rel="prev" title="Xfo" href="../maths/xfo.html"/> 

  
  <script src="_static/js/modernizr.min.js"></script>

</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-nav-search">
        

        
          <a href="../../../index.html" class="icon icon-home"> Kraken Rigging Framework
        

        
        </a>

        
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

        
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
        
          
          
              <ul>
<li class="toctree-l1"><a class="reference internal" href="../../../license.html">License</a><ul class="simple">
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../../installation.html">Installation</a><ul class="simple">
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../../releases.html">Releases</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../releases/release_1.0.0.html">Release 1.0.0</a><ul class="simple">
</ul>
</li>
</ul>
</li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../integrations.html">Integrations</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../integrations/maya.html">Maya Integration</a><ul class="simple">
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../../../integrations/softimage.html">Softimage Integration</a><ul class="simple">
</ul>
</li>
</ul>
</li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../user_interface.html">User Interface</a><ul class="simple">
</ul>
</li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../kl/kl_guide.html">KL Guide</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../kl/kl_library.html">KL Library</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../kl/solvers/kraken_solver.html">Kraken Solver</a><ul class="simple">
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../../../kl/solvers/kraken_arg.html">Kraken Arg</a><ul class="simple">
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../../../kl/solvers/bezier_spine_solver.html">Bezier Spine Solver</a><ul class="simple">
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../../../kl/solvers/kraken_multiposecon_solver.html">Kraken Multi-Pose Constraint Solver</a><ul class="simple">
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../../../kl/solvers/kraken_posecon_solver.html">Kraken Pose Constraint Solver</a><ul class="simple">
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../../../kl/solvers/kraken_rigscale_solver.html">Kraken Rig Scale Solver</a><ul class="simple">
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../../../kl/solvers/nbone_ik_solver.html">N-Bone IK Solver</a><ul class="simple">
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../../../kl/solvers/twobone_ik_solver.html">Two-Bone IK Solver</a><ul class="simple">
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../../../kl/tutorials/custom_solvers.html">Custom Solvers</a><ul class="simple">
</ul>
</li>
</ul>
</li>
</ul>
<ul class="current">
<li class="toctree-l1 current"><a class="reference internal" href="../../python_guide.html">Python Guide</a><ul class="current">
<li class="toctree-l2 current"><a class="reference internal" href="../../python_library.html">Python Library</a><ul class="current">
<li class="toctree-l3 current"><a class="reference internal" href="../core.html">Kraken Core</a><ul class="current">
<li class="toctree-l4"><a class="reference internal" href="../builder.html">Builder</a></li>
<li class="toctree-l4"><a class="reference internal" href="../kraken_system.html">Kraken System</a></li>
<li class="toctree-l4"><a class="reference internal" href="../profiler.html">Profiler</a></li>
<li class="toctree-l4"><a class="reference internal" href="../synchronizer.html">Synchronizer</a></li>
<li class="toctree-l4"><a class="reference internal" href="../configs/config.html">Configs</a></li>
<li class="toctree-l4"><a class="reference internal" href="../io/io.html">IO</a></li>
<li class="toctree-l4"><a class="reference internal" href="../maths/maths.html">Maths</a></li>
<li class="toctree-l4 current"><a class="current reference internal" href="">Objects</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../../helpers/helpers.html">Helpers</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../helpers/utility_methods.html">Utility Methods</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../../plugins/plugins.html">Plugins</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../plugins/maya_plugin/maya_plugin.html">Maya Plugin</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../plugins/si_plugin/si_plugin.html">Softimage Plugin</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../../ui/ui.html">Kraken UI</a><ul class="simple">
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../../tutorials/custom_components.html">Custom Components</a><ul class="simple">
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../../tutorials/scripted_builds.html">Scripted Builds</a><ul class="simple">
</ul>
</li>
</ul>
</li>
</ul>

          
        
      </div>
      &nbsp;
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
        <a href="../../../index.html">Kraken Rigging Framework</a>
      </nav>


      
      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href="../../../index.html">Docs</a> &raquo;</li>
      
          <li><a href="../../python_guide.html">Python Guide</a> &raquo;</li>
      
          <li><a href="../../python_library.html">Python Library</a> &raquo;</li>
      
          <li><a href="../core.html">Kraken Core</a> &raquo;</li>
      
    <li>Objects</li>
      <li class="wy-breadcrumbs-aside">
        
          
            <a href="../../../_sources/python/core/objects/objects.txt" rel="nofollow"> View page source</a>
          
        
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document">
            
  <div class="section" id="objects">
<h1>Objects<a class="headerlink" href="#objects" title="Permalink to this headline">¶</a></h1>
<p>This package represents all of the base objects that Kraken can create.</p>
<div class="section" id="modules">
<h2>Modules<a class="headerlink" href="#modules" title="Permalink to this headline">¶</a></h2>
<div class="toctree-wrapper compound">
<ul>
<li class="toctree-l1"><a class="reference internal" href="attributes/attributes.html">Attributes</a><ul>
<li class="toctree-l2"><a class="reference internal" href="attributes/attribute.html">Attribute</a></li>
<li class="toctree-l2"><a class="reference internal" href="attributes/attribute_group.html">Attribute Group</a></li>
<li class="toctree-l2"><a class="reference internal" href="attributes/bool_attribute.html">Boolean Attribute</a></li>
<li class="toctree-l2"><a class="reference internal" href="attributes/integer_attribute.html">Integer Attribute</a></li>
<li class="toctree-l2"><a class="reference internal" href="attributes/number_attribute.html">Number Attribute</a></li>
<li class="toctree-l2"><a class="reference internal" href="attributes/scalar_attribute.html">Scalar Attribute</a></li>
<li class="toctree-l2"><a class="reference internal" href="attributes/string_attribute.html">String Attribute</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="components/components.html">Components</a><ul>
<li class="toctree-l2"><a class="reference internal" href="components/component.html">Component</a></li>
<li class="toctree-l2"><a class="reference internal" href="components/base_example_component.html">Base Example Component</a></li>
<li class="toctree-l2"><a class="reference internal" href="components/component_input.html">Component Input</a></li>
<li class="toctree-l2"><a class="reference internal" href="components/component_input_port.html">Component Input Port</a></li>
<li class="toctree-l2"><a class="reference internal" href="components/component_output.html">Component Output</a></li>
<li class="toctree-l2"><a class="reference internal" href="components/component_output_port.html">Component Output Port</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="constraints/constraints.html">Constraints</a><ul>
<li class="toctree-l2"><a class="reference internal" href="constraints/constraint.html">Constraint</a></li>
<li class="toctree-l2"><a class="reference internal" href="constraints/orientation_constraint.html">Orientation Constraint</a></li>
<li class="toctree-l2"><a class="reference internal" href="constraints/pose_constraint.html">Pose Constraint</a></li>
<li class="toctree-l2"><a class="reference internal" href="constraints/position_constraint.html">Position Constraint</a></li>
<li class="toctree-l2"><a class="reference internal" href="constraints/scale_constraint.html">Scale Constraint</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="operators/operators.html">Operators</a><ul>
<li class="toctree-l2"><a class="reference internal" href="operators/operator.html">Operator</a></li>
<li class="toctree-l2"><a class="reference internal" href="operators/splice_operator.html">Splice Operator</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="component_group.html">Component Group</a><ul class="simple">
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="container.html">Container</a><ul class="simple">
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="control.html">Control</a><ul class="simple">
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="ctrlSpace.html">Control Space</a><ul class="simple">
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="curve.html">Curve</a><ul class="simple">
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="hierarchy_group.html">Hierarchy Group</a><ul class="simple">
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="joint.html">Joint</a><ul class="simple">
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="layer.html">Layer</a><ul class="simple">
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="locator.html">Locator</a><ul class="simple">
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="object_3d.html">Object 3D</a><ul class="simple">
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="rig.html">Rig</a><ul class="simple">
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="scene_item.html">Scene Item</a><ul class="simple">
</ul>
</li>
</ul>
</div>
</div>
<div class="section" id="module-kraken.core.objects">
<span id="module-methods"></span><h2>Module Methods<a class="headerlink" href="#module-kraken.core.objects" title="Permalink to this headline">¶</a></h2>
<p>Kraken Objects.</p>
<div class="line-block">
<div class="line"><br /></div>
<div class="line"><br /></div>
</div>
</div>
<div class="section" id="indices-and-tables">
<h2>Indices and tables<a class="headerlink" href="#indices-and-tables" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li><a class="reference internal" href="../../../genindex.html"><span>Index</span></a></li>
<li><a class="reference internal" href="../../../py-modindex.html"><span>Module Index</span></a></li>
<li><a class="reference internal" href="../../../search.html"><span>Search Page</span></a></li>
</ul>
<div class="line-block">
<div class="line"><br /></div>
</div>
</div>
</div>


          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="attributes/attributes.html" class="btn btn-neutral float-right" title="Attributes" accesskey="n">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="../maths/xfo.html" class="btn btn-neutral" title="Xfo" accesskey="p"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2015, Eric Thivierge, Phil Taylor, Hybride Technologies, Fabric Software Inc..
    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/snide/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>.

</footer>

        </div>
      </div>

    </section>

  </div>
  


  

    <script type="text/javascript">
        var DOCUMENTATION_OPTIONS = {
            URL_ROOT:'../../../',
            VERSION:'1.0.0',
            COLLAPSE_INDEX:false,
            FILE_SUFFIX:'.html',
            HAS_SOURCE:  true
        };
    </script>
      <script type="text/javascript" src="../../../_static/jquery.js"></script>
      <script type="text/javascript" src="../../../_static/underscore.js"></script>
      <script type="text/javascript" src="../../../_static/doctools.js"></script>

  

  
  
    <script type="text/javascript" src="../../../_static/js/theme.js"></script>
  

  
  
  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.StickyNav.enable();
      });
  </script>
   

</body>
</html>