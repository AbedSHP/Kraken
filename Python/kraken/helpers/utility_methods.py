

def logHierarchy(kObject):
    """Traverses the given Kraken hierarchy and logs the names of all the objects.

    Return:
    None

    """

    print kObject.getFullName()
    for i in xrange(kObject.getNumChildren()):
        child = kObject.getChildByIndex(i)
        logHierarchy(child)



from kraken.core.maths.math_object import MathObject
from kraken.core.maths import decodeValue

def __convertFromJSON(jsonData):

    if type(jsonData) is list:
        newList = []
        for item in jsonData:
            newList.append(__convertFromJSON(item))
        return newList
    elif type(jsonData) is dict:
        if '__mathObjectClass__' in jsonData.keys():
            return decodeValue(jsonData)
        for key, value in jsonData.iteritems():
            jsonData[key] = __convertFromJSON(value)
    return jsonData

def prepareToLoad(jsonData):

    return __convertFromJSON(jsonData)

def __convertToJSON(jsonData):

    if isinstance(jsonData, MathObject):
        return jsonData.jsonEncode()
    elif type(jsonData) is list:
        newList = []
        for item in jsonData:
            newList.append(__convertToJSON(item))
        return newList
    elif type(jsonData) is dict:
        for key, value in jsonData.iteritems():
            jsonData[key] = __convertToJSON(value)
    return jsonData

def prepareToSave(jsonData):
    return __convertToJSON(jsonData)