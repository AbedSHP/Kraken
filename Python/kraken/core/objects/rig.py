"""Kraken - objects.rig module.

Classes:
Rig -- Rig representation.

"""

from container import Container
from kraken.core.kraken_system import KrakenSystem
from kraken.core.profiler import Profiler
from kraken.core.objects.layer import Layer
import importlib
import json

class Rig(Container):
    """Rig object."""

    def __init__(self, name):
        super(Rig, self).__init__(name)


    def loadRigDefinitionFile(self, filepath):
        Profiler.getInstance().push("LoadRigDefinitionFile:" + filepath)

        if not os.path.exists(filepath):
            raise Exception("File not found:" + filepath)
        jsonData = json.load(str(open( referencefile ).read()))
        this.loadRigDefinition(jsonData)
        Profiler.getInstance().pop()

    def loadRigDefinition(self, jsonData):

        Profiler.getInstance().push("loadRigDefinition:" + self.getName())

        krakenSystem = KrakenSystem.getInstance()

        def loadLayers(layersData):
            for layerName in layersData:
                layer = Layer(layerName, parent=self)

        def loadComponents(componentsData):
            Profiler.getInstance().push("loadComponents")

            for componentData in componentsData:
                moduleName = '.'.join(componentData['class'].split('.')[:-1])
                className = componentData['class'].split('.').pop()
                if moduleName is not "":
                    importlib.import_module(moduleName)

                componentClass = krakenSystem.getComponentClass(className)
                if 'name' in componentData:
                    componentName = componentData['name']
                else:
                    componentName = str(componentClass.__name__)                    
                component = componentClass(componentName, parent=self, data=componentData)

            Profiler.getInstance().pop()

        def makeConnections(connectionsData):
            Profiler.getInstance().push("makeConnections")

            for connectionData in connectionsData:
                sourceComponentName, outputName = connectionData['source'].split('.')
                targetComponentName, inputName = connectionData['target'].split('.')

                sourceComponent = self.getChildByName(sourceComponentName)
                if sourceComponent is None:
                    raise Exception("Error making connection:" + connectionData['source'] + " -> " + connectionData['target']+". Source component not found:" + sourceComponent)
                targetComponent = self.getChildByName(targetComponentName)
                if targetComponent is None:
                    raise Exception("Error making connection:" + connectionData['source'] + " -> " + connectionData['target']+". Source component not found:" + targetComponent)
                outputPort = sourceComponent.getOutputByName(outputName)
                if outputPort is None:
                    raise Exception("Error making connection:" + connectionData['source'] + " -> " + connectionData['target']+". Output '" + outputName + "' not found on Component:" + sourceComponent.getName())
                inputPort = targetComponent.getInputByName(inputName)
                if inputPort is None:
                    raise Exception("Error making connection:" + connectionData['source'] + " -> " + connectionData['target']+". Input '" + inputName + "' not found on Component:" + targetComponent.getName())
                inputPort.setSource(outputPort.getTarget())

            Profiler.getInstance().pop()

        if 'layers' in jsonData:
            loadLayers(jsonData['layers'])
        else:
            raise Exception("A rig must define layers.")

        if 'components' in jsonData:
            loadComponents(jsonData['components'])

            if 'connections' in jsonData:
                makeConnections(jsonData['connections'])


        Profiler.getInstance().pop()