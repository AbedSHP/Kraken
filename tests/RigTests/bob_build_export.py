import json
import os

from kraken import plugins
from kraken.core.objects.rig import Rig
from kraken.examples.bob_guide_data import bob_guide_data
from kraken.core.profiler import Profiler
from kraken.helpers.utility_methods import logHierarchy


Profiler.getInstance().push("bob_build")

bobGuideRig = Rig("char_bob")
bobGuideRig.loadRigDefinition(bob_guide_data)

bobGuideRig.writeGuideDefinitionFile('bob_guide.rigdata')

bobGuideRig2 = Rig()
bobGuideRig2.loadRigDefinitionFile('bob_guide.rigdata')

logHierarchy(bobGuideRig2)

os.remove('bob_guide.rigdata')