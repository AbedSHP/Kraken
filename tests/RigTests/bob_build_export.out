char_bob
char_bob.spine:M
char_bob.controls
char_bob.controls.spine
char_bob.controls.spine.inputs
char_bob.controls.spine.inputs.mainSrt
char_bob.controls.spine.outputs
char_bob.controls.spine.outputs.cog
char_bob.controls.spine.outputs.spineBase
char_bob.controls.spine.outputs.pelvis
char_bob.controls.spine.outputs.spineEnd
char_bob.controls.spine.cogPosition
char_bob.controls.spine.spine01Position
char_bob.controls.spine.spine02Position
char_bob.controls.spine.spine03Position
char_bob.controls.spine.spine04Position
char_bob.controls.neck
char_bob.controls.neck.inputs
char_bob.controls.neck.inputs.neckBase
char_bob.controls.neck.outputs
char_bob.controls.neck.outputs.neck
char_bob.controls.neck.outputs.neckEnd
char_bob.controls.neck.neck
char_bob.controls.neck.neckEnd
char_bob.controls.head
char_bob.controls.head.inputs
char_bob.controls.head.inputs.headBase
char_bob.controls.head.outputs
char_bob.controls.head.outputs.head
char_bob.controls.head.outputs.jaw
char_bob.controls.head.outputs.eyeL
char_bob.controls.head.outputs.eyeR
char_bob.controls.head.head
char_bob.controls.head.headEnd
char_bob.controls.head.eyeLeft
char_bob.controls.head.eyeRight
char_bob.controls.head.jaw
char_bob.controls.Clavicle
char_bob.controls.Clavicle.inputs
char_bob.controls.Clavicle.inputs.spineEnd
char_bob.controls.Clavicle.outputs
char_bob.controls.Clavicle.outputs.clavicle
char_bob.controls.Clavicle.outputs.clavicleEnd
char_bob.controls.Clavicle.clavicle
char_bob.controls.Clavicle.clavicleUpV
char_bob.controls.Clavicle.clavicleEnd
char_bob.controls.Clavicle
char_bob.controls.Clavicle.inputs
char_bob.controls.Clavicle.inputs.spineEnd
char_bob.controls.Clavicle.outputs
char_bob.controls.Clavicle.outputs.clavicle
char_bob.controls.Clavicle.outputs.clavicleEnd
char_bob.controls.Clavicle.clavicle
char_bob.controls.Clavicle.clavicleUpV
char_bob.controls.Clavicle.clavicleEnd
char_bob.controls.Arm
char_bob.controls.Arm.inputs
char_bob.controls.Arm.inputs.globalSRT
char_bob.controls.Arm.inputs.clavicleEnd
char_bob.controls.Arm.outputs
char_bob.controls.Arm.outputs.bicep
char_bob.controls.Arm.outputs.forearm
char_bob.controls.Arm.outputs.armEndXfo
char_bob.controls.Arm.outputs.hand
char_bob.controls.Arm.bicepFK
char_bob.controls.Arm.forearmFK
char_bob.controls.Arm.wristFK
char_bob.controls.Arm.hand
char_bob.controls.Arm
char_bob.controls.Arm.inputs
char_bob.controls.Arm.inputs.globalSRT
char_bob.controls.Arm.inputs.clavicleEnd
char_bob.controls.Arm.outputs
char_bob.controls.Arm.outputs.bicep
char_bob.controls.Arm.outputs.forearm
char_bob.controls.Arm.outputs.armEndXfo
char_bob.controls.Arm.outputs.hand
char_bob.controls.Arm.bicepFK
char_bob.controls.Arm.forearmFK
char_bob.controls.Arm.wristFK
char_bob.controls.Arm.hand
char_bob.controls.Leg
char_bob.controls.Leg.inputs
char_bob.controls.Leg.inputs.globalSRT
char_bob.controls.Leg.inputs.pelvisInput
char_bob.controls.Leg.outputs
char_bob.controls.Leg.outputs.femur
char_bob.controls.Leg.outputs.shin
char_bob.controls.Leg.outputs.legEndXfo
char_bob.controls.Leg.outputs.foot
char_bob.controls.Leg.outputs.toe
char_bob.controls.Leg.femur
char_bob.controls.Leg.knee
char_bob.controls.Leg.ankle
char_bob.controls.Leg.toe
char_bob.controls.Leg.toeTip
char_bob.controls.Leg
char_bob.controls.Leg.inputs
char_bob.controls.Leg.inputs.globalSRT
char_bob.controls.Leg.inputs.pelvisInput
char_bob.controls.Leg.outputs
char_bob.controls.Leg.outputs.femur
char_bob.controls.Leg.outputs.shin
char_bob.controls.Leg.outputs.legEndXfo
char_bob.controls.Leg.outputs.foot
char_bob.controls.Leg.outputs.toe
char_bob.controls.Leg.femur
char_bob.controls.Leg.knee
char_bob.controls.Leg.ankle
char_bob.controls.Leg.toe
char_bob.controls.Leg.toeTip
char_bob.neck:M
char_bob.head:M
char_bob.Clavicle:L
char_bob.Clavicle:R
char_bob.Arm:L
char_bob.Arm:R
char_bob.Leg:L
char_bob.Leg:R
